contrib/idn/mdnkit/tools/mdnconv/mdnconv.1
.\" $Id: mdnconv.1,v 1.17 2001/05/18 04:16:27 ishisone Exp $
.\"
.\" Copyright (c) 2000,2001 Japan Network Information Center.
.\" All rights reserved.
.\"  
.\" By using this file, you agree to the terms and conditions set forth bellow.
.\" 
.\" 			LICENSE TERMS AND CONDITIONS 
.\" 
.\" The following License Terms and Conditions apply, unless a different
.\" license is obtained from Japan Network Information Center ("JPNIC"),
.\" a Japanese association, Fuundo Bldg., 1-2 Kanda Ogawamachi, Chiyoda-ku,
.\" Tokyo, Japan.
.\" 
.\" 1. Use, Modification and Redistribution (including distribution of any
.\"    modified or derived work) in source and/or binary forms is permitted
.\"    under this License Terms and Conditions.
.\" 
.\" 2. Redistribution of source code must retain the copyright notices as they
.\"    appear in each source code file, this License Terms and Conditions.
.\" 
.\" 3. Redistribution in binary form must reproduce the Copyright Notice,
.\"    this License Terms and Conditions, in the documentation and/or other
.\"    materials provided with the distribution.  For the purposes of binary
.\"    distribution the "Copyright Notice" refers to the following language:
.\"    "Copyright (c) Japan Network Information Center.  All rights reserved."
.\" 
.\" 4. Neither the name of JPNIC may be used to endorse or promote products
.\"    derived from this Software without specific prior written approval of
.\"    JPNIC.
.\" 
.\" 5. Disclaimer/Limitation of Liability: THIS SOFTWARE IS PROVIDED BY JPNIC
.\"    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\"    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
.\"    PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL JPNIC BE LIABLE
.\"    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\"    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\"    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\"    BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\"    WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\"    OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
.\"    ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
.\" 
.\" 6. Indemnification by Licensee
.\"    Any person or entities using and/or redistributing this Software under
.\"    this License Terms and Conditions shall defend indemnify and hold
.\"    harmless JPNIC from and against any and all judgements damages,
.\"    expenses, settlement liabilities, cost and other liabilities of any
.\"    kind as a result of use and redistribution of this Software or any
.\"    claim, suite, action, litigation or proceeding by any third party
.\"    arising out of or relates to this License Terms and Conditions.
.\" 
.\" 7. Governing Law, Jurisdiction and Venue
.\"    This License Terms and Conditions shall be governed by and and
.\"    construed in accordance with the law of Japan. Any person or entities
.\"    using and/or redistributing this Software under this License Terms and
.\"    Conditions hereby agrees and consent to the personal and exclusive
.\"    jurisdiction and venue of Tokyo District Court of Japan.
.\"
.TH MDNCONV 1 "Mar 3, 2001"
.\"
.SH NAME
mdnconv \- codeset converter for named.conf and zone master files
.\"
.SH SYNOPSIS
\fBmdnconv\fP [\fIoptions..\fP] [\fIfile\fP...]
.\"
.SH DESCRIPTION
.B mdnconv
is a codeset converter for named configuration files and zone master files.
.B mdnconv
performs codeset conversion specified either by the command-line arguments
or by the configuration file,
and writes the converted text to stdout.
.PP
If file name is specified,
.B mdnconv
converts the contents of the file.  Otherwise,
.B mdnconv
converts
.IR stdin .
.PP
Since
.B mdnconv
is specifically designed for converting internatinalized domain names,
it may not be suitable as a general codeset converter.
.\"
.SH "OPERATION MODES"
.B mdnconv
has two operation modes.
.PP
One is a mode to convert local-encoded domain names to IDN-encoded
one.  Usually this mode is used for preparing domain names to be
listed in named configuration files or zone master files.
In this mode, the following processes are performed in addition to
the codeset (encoding) conversion.
.RS 2
.IP \- 2
local mapping
.IP \- 2
standard domain name preperation (NAMEPREP)
.RE
.PP
Furthermore, delimiter mapping is also performed in case \-delimitermap
command line option is specified.
.PP
The other mode is a reverse convertion, from IDN-encoded domain name to
local-encoded domain names.
Since IDN-ecoded names are usually already normalized by local mapping and
NAMEPREP process, only codeset conversion is performed in this mode.
.\"
.SH OPTIONS
Normally
.B mdnconv
reads system's default configuration file (mdn.conf) and perform
conversion or name preparation according to the parameters specified in
the file.  You can override the setting in the configuration file by
various command line options below.
.TP 4
\fB\-in\fP \fIin-code\fP, \fB\-i\fP \fIin-code\fP
Specify the codeset name of the input text.
Any of the following codeset names can be specified.
.RS 4
.IP "\(bu" 2
Any codeset names which \fIiconv_open()\fP library function accepts
.IP "\(bu" 2
\f(CWRACE\fR
.IP "\(bu" 2
\f(CWBRACE\fR
.IP "\(bu" 2
\f(CWLACE\fR
.IP "\(bu" 2
\f(CWDUDE\fR
.IP "\(bu" 2
\f(CWUTF-5\fR
.IP "\(bu" 2
\f(CWUTF-6\fR
.IP "\(bu" 2
\f(CWAltDUDE\fR
.IP "\(bu" 2
\f(CWAMC-ACE-O\fR
.IP "\(bu" 2
\f(CWAMC-ACE-M\fR
.IP "\(bu" 2
\f(CWAMC-ACE-R\fR
.IP "\(bu" 2
Any alias names for the above, defined by the codeset alias file.
.RE
.IP "" 4
If this option is not specified, the default codeset is determined
from the locale in normal conversion mode.
In reverse conversion mode, the default codeset is the IDN encoding
specified by the configuration file (``idn-encoding'' entry).
.TP 4
\fB\-out\fP \fIout-code\fP, \fB\-o\fP \fIout-code\fP
Specify the codeset name of the output text. \fIout-code\fP can be any
codeset name that can be specified for 
.B \-in
option.
.IP "" 4
If this option is not specified, the default is the IDN encoding
specified by the configuration file (``idn-encoding'' entry) in
normal conversion mode.
In reverse conversion mode, the default codeset is determined from
the locale.
.TP 4
\fB\-conf\fP \fIpath\fP, \fB\-c\fP \fIpath\fP
Specify the pathname of mDNkit configuration file (``mdn.conf'').
If not specified, system's default file is used, unless \-noconf
option is specified.
.TP 4
\fB\-noconf\fP, \fB\-C\fP
Specify that no configuration file is to be used.
.TP 4
\fB\-reverse\fP, \fB\-r\fP
Specify reverse conversion mode.
.br
If this option is not specified, the normal conversion mode is used.
.TP 4
\fB\-nameprep\fR \fIversion\fR, \fB\-n\fR \fIversion\fR
Specify the version of NAMEPREP.
The following is a list of currently available
.IR version s.
.RS 4
.IP \f(CWnameprep-02\fR 4
Perform NAMEPREP according to the Internet Draft
``draft-ietf-idn-nameprep-02.txt'' (NAMEPREP-02).
.IP \f(CWnameprep-03\fR 4
Perform NAMEPREP according to the Internet Draft
``draft-ietf-idn-nameprep-03.txt'' (NAMEPREP-03).
.RE
.IP "" 4
This option is a shorthand for specifying the 4 options below.
.PP
.RS 8
.ft CW
-map 
.I version
.br
-normalize
.I version
.br
-prohibit
.I version
.br
-unassigned
.I version
.ft R
.RE
.IP "" 4
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-nonameprep\fR, \fB\-N\fR
Specify to skip NAMEPREP process in the normal conversion mode.
.br
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-map\fR \fImap\fR
Specify the name of the mapping rule for NAMEPREP mapping process.
Currently, following
.IR map s
are available.
.RS 4
.IP \f(CWnameprep-02\fR 4
Use the list of mappings specified by NAMEPREP-02 draft.
.IP \f(CWnameprep-03\fR 4
Use the list of mappings specified by NAMEPREP-03 draft.
.IP \f(CWfilemap:\fR\fIpath\fR 4
Use list of mappings specified by mapfile
.IR path .
See mdn.conf(5) for the format of a mapfile.
.RE
.IP "" 4
This option can be specified more than once.
In that case, each mapping will be performed in the order of the
specification.
.br
If neither this option nor ``\-nameprep'' are specified,
the schemes specified in the configuration file (``nameprep-map'' entry)
are used.
.br
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-normalize\fP \fIscheme\fP
Specify the normalization scheme for NAMEPREP normalization process.
Currently, following \fIscheme\fPs are available:
.RS 4
.IP \f(CWascii-uppercase\fR
ASCII lowercase to uppercase mapping.
.IP \f(CWascii-lowercase\fR
ASCII uppercase to lowercase mapping.
.IP \f(CWunicode-uppercase\fR
Unicode lowercase to uppercase mapping, which is described in
``Unicode Technical Report #21: Case Mappings'',
available from unicode.org.
.IP \f(CWunicode-lowercase\fR
Unicode uppercase to lowercase mapping, according ``Case Mappings''.
.IP \f(CWunicode-foldcase\fR
Unicode case-folding for case-less string matching,
also described in the above document.
.IP \f(CWunicode-form-c\fR
Unicode Normlization Form C, which is described in
``Unicode Technical Report #15: Unicode Normalization Forms'',
also available from unicode.org.
.IP \f(CWunicode-form-kc\fR
Unicode Normlization Form KC, also described in the above
document.
.IP \f(CWnameprep-02\fR
Synonim for ``unicode-form-kc''.
.IP \f(CWnameprep-03\fR
Synonim for ``unicode-form-kc''.
.RE
.IP "" 4
This option can be specified more than once.
In that case, each normalization will be performed in the order of the
specification.
.br
If neither this option nor ``\-nameprep'' are specified,
the schemes specified in the configuration file (``nameprep-normalize'' entry)
are used.
.br
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-prohibit\fR \fIset\fR
Specify the set of prohibited characters.
Currently, following
.IR set s
are available.
.RS 4
.IP \f(CWnameprep-02\fR 4
Use list of prohibit characters specified by NAMEPREP-02 draft.
.IP \f(CWnameprep-03\fR 4
Use list of prohibit characters specified by NAMEPREP-03 draft.
.IP \f(CWfileset:\fR\fIpath\fR 4
Use list of characters specified by setfile
.IR path .
See mdn.conf(5) for the format of a setfile.
.RE
.IP "" 4
This option can be specified more than once.  In this case, the set of
prohibited characters is the union of the specified sets.
.br
If neither this option nor ``\-nameprep'' is specified,
list of characters specified by the configuration file
(``nameprep-prohibit'' entry) is used.
.br
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-unassigned\fR \fIset\fR
Specify the set of unassigned codepoints.
Currently, following
.IR set s
are available.
.RS 4
.IP \f(CWnameprep-02\fR 4
Use the list of unassigned characters specified by NAMEPREP-02 draft.
.IP \f(CWnameprep-03\fR 4
Use the list of unassigned characters specified by NAMEPREP-03 draft.
.IP \f(CWfileset:\fR\fIpath\fR 4
Use list of characters specified by setfile
.IR path .
See mdn.conf(5) for the format of a setfile.
.RE
.IP "" 4
Note that unassigned codepoint checking is not performed
unless \-unassigncheck option is specified.
.br
This option can be specified more than once.  In this case, the set of
prohibited characters is the union of the specified sets.
.br
If neither this option nor ``\-nameprep'' is specified,
list of characters specified by the configuration file
(``nameprep-unassigned'' entry) is used.
.br
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-unassigncheck\fR, \fB\-u\fR
Specify unassigned codepoint check should be performed.
Unassigned codepoint check is disabled by default.
.br
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-delimiter\fR \fIcodepoint\fP
Specify the character to be mapped to domain name delimiter (period).
This option can be specified more than once in order to specify multiple
characters.
.br
If this option is not specified, characters specified by the
configuration file (``delimiter-map'') is used.
.br
Note that delimiter mapping is not performed unless \-delimitermap option
is specified.
.br
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-delimitermap\fR, \fB\-d\fR
Specify delimiter mapping should be performed.
By default, delimiter mapping is disabled.
.br
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-whole\fP, \fB\-w\fP
Perform local mapping, nameprep and conversion to output codeset for the entire
input text.  If this option is not specified, only non-ASCII characters
and their surrounding texts will be processed.
See ``CONVERSION MECHANISM'' for details.
.TP 4
\fB\-alias\fP \fIpath\fP, \fB\-a\fP \fIpath\fP
Specify a codeset alias file.  It is a simple text file, where
each line has a pair of alias name and real name separated by one
or more white spaces like below:
.nf
.ft CW

    \fIalias-codeset-name\fP    \fIreal-codeset-name\fP

.ft R
.fi
Lines starting with ``#'' are treated as comments.
.TP 4
\fB\-flush\fP
Force line-buffering mode.
.TP 4
\fB\-version\fP, \fB\-v\fP
Print version information and quit.
.\"
.SH CONVERSION MECHANISM
Here is a brief step-by-step description of how
.B mdnconv
does its job.
.IP "1. input-side convnersion" 4
.B mdnconv
reads input text line by line, and converts them from the input codeset
to UTF-8 encoding.  Basically this conversion is done using iconv() utility,
but for some special codesets like RACE, special converter will be used.
.br
Normally the whole line is converted, but if the input codeset is an
ACE (ASCII-compatible encoding, such as RACE), a special conversion
prodedure shown below is used:
.RS 4
.IP 1) 3
Find substrings which are valid ASCII domain names (i.e. constructed from
alphanumeric characters and hyphens) in the line.
.IP 2) 3
For each substring found, try converting it to UTF-8.
.IP 3) 3
If the conversion succeeds, replace the substring with the conversion
result. Otherwise, the original substring is retained.
.RE
.IP "2. finding where to process" 4
Then
.B mdnconv
looks for (part of) internationalized domain name in the converted line,
and picks them up for further processing.
.br
.B mdnconv
recognizes the character sequence having the following property as
an internationalized domain name.
.RS 4
.IP "\(bu" 2
containing at least one non-ASCII character, and
.IP "\(bu" 2
consisting of legal domain name characters (alphabets, digits, hypens),
non-ASCII characters and period.
.RE
.IP "" 4
If you are not quite sure what this means (as I), 
the following regular expression may help you.
.nf
.ft CW

    [-A-Za-z0-9.]*<non-ASCII char>([-A-Za-z0-9.]|<non-ASCII char>)*

.ft P
.fi
For example, only ``www.###.com'' will be picked up from
the following lines (`#' denotes non-ASCII character).
.nf
.ft CW

    www.###.com.    IN CNAME serv.xyz.com.
    serv.xyz.com.   IN A     10.0.0.1

.ft P
.fi
If \-whole option is specified, this step is skipped and the entire line
is subject to further processing.
.IP "3. delimiter mapping" 4
For each string picked up by the last step, delimiter mapping is performed.
That is, substibute certain characters specified as domain name delimiter
with period.
.IP "4. local mapping" 4
Perform local mapping.
If the local mapping is specified by command line option ``\-localmap'',
the specified mapping rule is applied.  Otherwise, find the mapping rule
from the configuration file which matches to the TLD of the name,
and perform mapping according to the matched rule.
.IP "5. nameprep" 4
Perform name preparation (NAMEPREP).  This is a following 3-step process.
.RS 6
.nf
1) Mapping
2) Normalization
3) Prohibited/unassigned character checking
.fi
.RE
.IP "6. output-side conversion" 4
Finally,
.B mdnconv
converts the nameprepped strings from UTF-8 encoding to the output codeset.
And it writes them to stdout together with the rest of the line.
.PP
If
.B mdnconv
operates in the reverse conversion mode, step 3, 4 and 5 above are skipped.
.\"
.SH FILE MANAGEMENT
Maybe the best way to manage named.conf or zone master files that contains
internationalized domain name is to keep them in your local codeset so that
they can be edited with your favorite editor, and generate a version in
the IDN encoding using \fBmdnconv\fP.
.PP
`make' is a convenient tool for this purpose.
Suppose the local codeset version has suffix `.lc', and its ACE version
has suffix `.ace'.  The following Makefile enables you to generate
ACE version from local codeset version by just typing `make'.
.RS 4
.nf
.ft CW

\&.SUFFIXES: .lc .ace
\&.lc.ace:
        mdnconv -in $(LOCALCODE) $(MDNCONVOPT) $< > $@

LOCALCODE = EUC-JP
MDNCONVOPT = 

DESTFILES = db.zone1.ace db.zone2.ace

all: $(DESTFILES)
.ft
.fi
.RE
.\"
.SH SEE ALSO
mdn.conf(5),
iconv(3)
.\"
.SH BUGS
The automatic input-code selection depends on your system, and sometimes
it cannot guess or guess wrong.  It is better to explicitly specify it
using \-in option.


NEWS
doc/arm/changelog.rst

contrib/idn/mdnkit/tools/mdnconv/mdnconv.c
#ifndef lint
static char *rcsid = "$Id: mdnconv.c,v 1.19 2000/09/13 05:55:55 ishisone Exp $";
#endif

/*
 * Copyright (c) 2000 Japan Network Information Center.  All rights reserved.
 *  
 * By using this file, you agree to the terms and conditions set forth bellow.
 * 
 * 			LICENSE TERMS AND CONDITIONS 
 * 
 * The following License Terms and Conditions apply, unless a different
 * license is obtained from Japan Network Information Center ("JPNIC"),
 * a Japanese association, Fuundo Bldg., 1-2 Kanda Ogawamachi, Chiyoda-ku,
 * Tokyo, Japan.
 * 
 * 1. Use, Modification and Redistribution (including distribution of any
 *    modified or derived work) in source and/or binary forms is permitted
 *    under this License Terms and Conditions.
 * 
 * 2. Redistribution of source code must retain the copyright notices as they
 *    appear in each source code file, this License Terms and Conditions.
 * 
 * 3. Redistribution in binary form must reproduce the Copyright Notice,
 *    this License Terms and Conditions, in the documentation and/or other
 *    materials provided with the distribution.  For the purposes of binary
 *    distribution the "Copyright Notice" refers to the following language:
 *    "Copyright (c) Japan Network Information Center.  All rights reserved."
 * 
 * 4. Neither the name of JPNIC may be used to endorse or promote products
 *    derived from this Software without specific prior written approval of
 *    JPNIC.
 * 
 * 5. Disclaimer/Limitation of Liability: THIS SOFTWARE IS PROVIDED BY JPNIC
 *    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 *    PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL JPNIC BE LIABLE
 *    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 *    BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 *    WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 *    OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 *    ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
 * 
 * 6. Indemnification by Licensee
 *    Any person or entities using and/or redistributing this Software under
 *    this License Terms and Conditions shall defend indemnify and hold
 *    harmless JPNIC from and against any and all judgements damages,
 *    expenses, settlement liabilities, cost and other liabilities of any
 *    kind as a result of use and redistribution of this Software or any
 *    claim, suite, action, litigation or proceeding by any third party
 *    arising out of or relates to this License Terms and Conditions.
 * 
 * 7. Governing Law, Jurisdiction and Venue
 *    This License Terms and Conditions shall be governed by and and
 *    construed in accordance with the law of Japan. Any person or entities
 *    using and/or redistributing this Software under this License Terms and
 *    Conditions hereby agrees and consent to the personal and exclusive
 *    jurisdiction and venue of Tokyo District Court of Japan.
 */

/*
 * mdnconv -- Codeset converter for named.conf and zone files
 */

#include <config.h>

#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <errno.h>
#ifdef HAVE_LOCALE_H
#include <locale.h>
#endif

#include <mdn/result.h>
#include <mdn/converter.h>
#include <mdn/normalizer.h>
#include <mdn/localencoding.h>
#include <mdn/utf8.h>
#include <mdn/resconf.h>

#include "util.h"

/* Maxmum number of normalizers */
#define MAX_NORMALIZER	10

#ifndef MDN_UTF8_ENCODING_NAME
#define MDN_UTF8_ENCODING_NAME "UTF-8"
#endif

#ifndef MDN_RACE_ENCODING_NAME
#define MDN_RACE_ENCODING_NAME "RACE"
#endif

#ifndef MDN_RACE_PREFIX
#define MDN_RACE_PREFIX "bq--"
#endif

int			line_number;	/* current input file line number */
int			flush_every_line = 0; /* pretty obvious */
mdn_converter_t		conv_in_ctx;	/* input converter */
mdn_converter_t		conv_out_ctx;	/* output converter */
mdn_normalizer_t	norm_ctx;	/* normalizer */

void			errormsg(const char *fmt, ...);
static int		convert_file(FILE *fp, char *zld, int auto_zld,
				     int selective);
static void		usage(char *cmd);

int
main(int ac, char **av) {
	char *cmd = *av;
	char *normalizer[MAX_NORMALIZER];
	int nnormalizer = 0;
	const char *in_code = NULL;
	const char *out_code = NULL;
	char *resconf_file = NULL;
	int no_resconf = 0;
	char zld[256 + 1];
	int zld_specified = 0;
	int auto_zld = 0;
	char *encoding_alias = NULL;
	int selective = 1;
	FILE *fp;
	mdn_resconf_t resconf;

#ifdef HAVE_SETLOCALE
	(void)setlocale(LC_ALL, "");
#endif

	zld[0] = '\0';

	ac--;
	av++;
	while (ac > 0 && **av == '-') {

#define MUST_HAVE_ARG if (ac < 2) usage(cmd)
		if (strcmp(*av, "-in") == 0) {
			MUST_HAVE_ARG;
			in_code = av[1];
			ac--;
			av++;
		} else if (strcmp(*av, "-out") == 0) {
			MUST_HAVE_ARG;
			out_code = av[1];
			ac--;
			av++;
		} else if (strcmp(*av, "-conf") == 0) {
			MUST_HAVE_ARG;
			resconf_file = av[1];
			ac--;
			av++;
		} else if (strcmp(*av, "-noconf") == 0) {
			no_resconf = 1;
		} else if (strcmp(*av, "-zld") == 0) {
			MUST_HAVE_ARG;
			canonical_zld(zld, av[1]);
			zld_specified = 1;
			ac--;
			av++;
		} else if (strcmp(*av, "-auto") == 0) {
			auto_zld = 1;
		} else if (strcmp(*av, "-normalize") == 0) {
			MUST_HAVE_ARG;
			if (nnormalizer >= MAX_NORMALIZER) {
				errormsg("too many normalizers\n");
				exit(1);
			}
			normalizer[nnormalizer++] = av[1];
			ac--;
			av++;
		} else if (strcmp(*av, "-alias") == 0) {
			MUST_HAVE_ARG;
			encoding_alias = *av;
		} else if (strcmp(*av, "-flush") == 0) {
			flush_every_line = 1;
		} else if (strcmp(*av, "-whole") == 0) {
			selective = 0;
		} else {
			usage(cmd);
		}
#undef MUST_HAVE_ARG

		ac--;
		av++;
	}

	if (ac > 1)
		usage(cmd);

	/*
	 * Load configuration file.
	 */
	resconf = NULL;
	if (!no_resconf) {
		mdn_result_t r;

		r = mdn_resconf_initialize();
		if (r == mdn_success)
			r = mdn_resconf_create(&resconf);
		if (r == mdn_success)
			r = mdn_resconf_loadfile(resconf, resconf_file);
		if (r != mdn_success) {
			errormsg("error reading configuration file: %s\n",
				 mdn_result_tostring(r));
			return (1);
		}
	}

	/*
	 * Get default input/output code.
	 */
	if (in_code == NULL)
		in_code = mdn_localencoding_name();

	if (out_code == NULL) {
		mdn_converter_t c;
		if (resconf != NULL &&
		    (c = mdn_resconf_serverconverter(resconf)) != NULL)
			out_code = mdn_converter_localencoding(c);
	}

	if (in_code == NULL) {
		errormsg("input codeset must be specified\n");
		return (1);
	}
	if (out_code == NULL) {
		errormsg("output codeset must be specified\n");
		return (1);
	}

	/*
	 * Initialize codeset converter.
	 */
	if (!initialize_converter(in_code, out_code, encoding_alias))
		return (1);

	/*
	 * Initialize normalizer.
	 */
	if (nnormalizer == 0 && resconf != NULL)
		norm_ctx = mdn_resconf_normalizer(resconf);
	if (norm_ctx == NULL &&
	    !initialize_normalizer(normalizer, nnormalizer))
		return (1);

	/*
	 * Default ZLD.
	 */
	if (!zld_specified && resconf != NULL) {
		const char *conf_zld = mdn_resconf_zld(resconf);
		if (conf_zld != NULL)
			canonical_zld(zld, conf_zld);
	}

	/*
	 * Open input file.
	 */
	if (ac > 0) {
		if ((fp = fopen(av[0], "r")) == NULL) {
			errormsg("cannot open file %s: %s\n",
				 av[0], strerror(errno));
			return (1);
		}
	} else {
		fp = stdin;
	}

	/*
	 * Do the conversion.
	 */
	return convert_file(fp, zld, auto_zld, selective);
}

static int
convert_file(FILE *fp, char *zld, int auto_zld, int selective) {
	mdn_result_t r;
	char line1[1024];
	char line2[1024];
	int nl_trimmed;
	int race_hack;

	if (strcmp(mdn_converter_localencoding(conv_in_ctx),
		   MDN_RACE_ENCODING_NAME) == 0)
		race_hack = 1;
	else
		race_hack = 0;

	line_number = 1;
	while (fgets(line1, sizeof(line1), fp) != NULL) {
		/*
		 * Trim newline at the end.  This is needed for
		 * those ascii-comatible encodings such as UTF-5 or RACE
		 * not to try converting newlines, which will result
		 * in `invalid encoding' error.
		 */
		if (line1[strlen(line1) - 1] == '\n') {
			line1[strlen(line1) - 1] = '\0';
			nl_trimmed = 1;
		} else {
			nl_trimmed = 0;
		}

		/*
		 * Convert input line to UTF-8.
		 */
		if (race_hack) {
			/*
			 * Find portions encoded in RACE using RACE's prefix
			 * "bq--", and selectively decode those portions.
			 */
			r = selective_decode(MDN_RACE_PREFIX,
					     line1, line2, 1024);
		} else {
			r = mdn_converter_localtoutf8(conv_in_ctx,
						      line1, line2, 1024);
		}
		if (r != mdn_success) {
			errormsg("conversion failed at line %d: %s\n",
				 line_number,
				 mdn_result_tostring(r));
			return (1);
		}
		if (!mdn_utf8_isvalidstring(line2)) {
			errormsg("conversion to utf-8 failed at line %d\n",
				 line_number);
			return (1);
		}

		/*
		 * Normalize and convert to the output codeset.
		 */
		if (selective) {
			r = selective_encode(line2, line1, sizeof(line1),
					     zld, auto_zld);
		} else {
			r = encode_region(line2, line1, sizeof(line1),
					  zld, auto_zld);
		}
		if (r != mdn_success)
			return (1);

		fputs(line1, stdout);
		if (nl_trimmed)
			putc('\n', stdout);

		if (flush_every_line)
			fflush(stdout);

		line_number++;
	}
	return (0);
}

void
errormsg(const char *fmt, ...) {
	va_list args;

	va_start(args, fmt);
	vfprintf(stderr, fmt, args);
	va_end(args);
}

static void
usage(char *cmd) {
	int i;
	static char *options[] = {
		"-in input-codeset   : specifies input codeset name.",
		"-out output-codeset : specifies output codeset name.",
		"-normalize scheme   : specifies normalization scheme.",
		"                      this option can be specified",
		"                      multiple times.",
		"-zld zld            : specifies ZLD to use.",
		"-auto               : automatically appends ZLD where",
		"                      seemed appropriate.",
		"-alias alias-file   : specifies codeset alias file.",
		"-conf conf-file     : specifies pathname of MDN resolver",
		"                      configuration file.",
		"-noconf             : do not load MDN resolver configuration",
		"                      file.",
		"-flush              : line-buffering mode.",
		"-whole              : convert the whole region instead of",
		"                      regions containing non-ascii characters",
		NULL,
	};

	fprintf(stderr, "Usage: %s [options..] [file]\n", cmd);

	for (i = 0; options[i] != NULL; i++)
		fprintf(stderr, "\t%s\n", options[i]);

	exit(1);
}
